src/core/document.js

```diff
@@ -787,11 +787,11 @@ class PDFDocument {
   get numPages() {
     let num = 0;
     if (this.xfaFactory) {
-      num = this.xfaFactory.numPages;
+      num = this.xfaFactory.getNumPages();
     } else if (this.linearization) {
-      num = this.linearization.numPages;
+      num = Promise.resolve(this.linearization.numPages);
     } else {
-      num = this.catalog.numPages;
+      num = Promise.resolve(this.catalog.numPages);
```

Comment:
I don't think that the way you're using Promises here (and elsewhere) in combination with the `PDFManager.ensure...`-method usage, is correct and/or safe in general.
For one thing, the `this.catalog.numPages` getter can throw and I don't think that `Promise.resolve(/* previously thrown error */)` makes a lot of sense.
Furthermore, since the `PDFManager.ensure...`-methods only goes one level "deep" so to speak, I'm not sure that the associated `ensure`-call in `src/core/worker.js` does what it *should* now.

This should be fixable with a bunch more re-factoring, and I can look into that and try to provide more actionable review comments during the weekend; however for now I figured that just pointing these things out makes sense (such that this won't land as-is).